src/sphinx_autodoc_typehints/__init__.py

```diff
@@ -214,6 +214,13 @@ def format_annotation(annotation: Any, config: Config) -> str:  # noqa: C901, PL
     args_format = "\\[{}]"
     formatted_args: str | None = ""
 
+    always_use_bars_union: bool = getattr(config, "always_use_bars_union", False)
```

Comment:
@hoodmane this is fantastic, thank you!

I would suggest that this should be the *only* behavior at this point in time and that the config option is unnecessary. Python itself encourages `|` over `Optional` and it therefore seems unnecessary to support an old and discouraged alternative. If the sphinx-autodoc-typehints maintainers believe it necessary to continue supporting the old way, then at a *minimum* the `always_use_bars_union` should default True. @flying-sheep any other comments?

Reply:
> it therefore seems unnecessary to support an old and discouraged alternative. If the sphinx-autodoc-typehints maintainers believe it necessary to continue supporting the old way, then at a minimum the always_use_bars_union should default True.

I will leave this to @gaborbernat to decide. But at least this gives a way to opt into the preferred behavior, even if the discoverability isn't necessarily the best.

Reply:
Either is opt-in or this will need to be a major release as is in a sense backwards incompatible change.

Reply:
In that case, is there anything speaking against making this the default (Or not making it a setting at all and just changing the rendering)?

Reply:
My 2c, let's keep the flag and make it a major release. That seems to be a solution "most" would be ok with.

Reply:
> keep the flag and make it a major release

Do you mean additionally to set the default value of the flag to `True`? Since if the default value is `False` only a minor release would be needed?

Reply:
I meant keep the flag, set it to True and make it a major release. The CI needs fixing too 😊